### PR TITLE
fix: select gas account for private send when megafuel is suppressed (OK-59993)

### PR DESCRIPTION
**Pull request by weatherstar**
[OK-59993](https://onekeyhq.atlassian.net/browse/OK-59993)

----------

## Summary

Follow-up to #12867, addressing the P1 finding from the automated review: https://github.com/OneKeyHQ/app-monorepo/pull/12867#issuecomment-5322816243

Private Send suppresses Megafuel, but `selectedPayer` was still derived from the raw backend `payer`. On BNB-chain responses where the fee service returns `payer='megafuel'` together with an eligible Gas Account quote, the confirm flow cleared Megafuel for display yet kept `selectedPayer='user'`, so the sponsored quote was silently dropped and the tx broadcast user-paid.

## Fix

- Extract the payer derivation into `resolveSponsorPayerState()` (`packages/kit/src/views/SignatureConfirm/utils/gasAccountPayerSelection.ts`), which derives `effectiveFeePayer` (display) and `selectedPayer` (submit wiring) together from the post-filtered sponsor state so they can no longer drift.
- When Megafuel is suppressed for Private Send and the server prefers `megafuel`, an eligible Gas Account quote now falls through to `selectedPayer='gasAccount'` with its `idempotencyKey`, and `effectiveFeePayer` follows — the sponsored coupon shows and the broadcast uses the sponsored submit path (`TxConfirmActions` requires both payers aligned).
- All other combinations (custom RPC, batch, scenario-disabled, temporarily-disabled, megafuel actually available) keep their existing behavior.

## Tests

- 16 new unit tests in `gasAccountPayerSelection.test.ts` covering the preserved behaviors and the Private Send fallback.
- `yarn agent:check --profile pr` passes.

[OK-59993](https://onekeyhq.atlassian.net/browse/OK-59993)

[OK-59993]: https://onekeyhq.atlassian.net/browse/OK-59993?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by sidmorizon** [comment]
@codex review

**Comment by sidmorizon** [comment]
@codex security review

**Comment by sidmorizon** [comment]
@cursoragent review

File: packages/kit/src/views/SignatureConfirm/components/TxFee/TxFeeInfo.tsx [hunk 1]
```diff
@@ -560,15 +561,28 @@ function TxFeeInfo(props: IProps) {
...
-          (gasAccountTemporarilyDisabled && serverPayer === 'gasAccount')
-            ? 'user'
-            : serverPayer;
+        const {
+          effectiveFeePayer: nextEffectiveFeePayer,
+          selectedPayer: nextSelectedPayer,
+        } = resolveSponsorPayerState({
+          serverPayer,
+          megafuelSponsorable: !!r.megafuelEligible?.sponsorable,
+          gasAccountQuoteEligible: !!(
```
Comment (sidmorizon) [inline packages/kit/src/views/SignatureConfirm/components/TxFee/TxFeeInfo.tsx]: # 🟠 P1 gasAccountQuoteEligible 未校验 quoteId

## 问题

`gasAccountQuoteEligible` 只判断 `r.gasAccountEligible && r.gasAccountQuote` 对象是否存在，没有校验 `quoteId` 非空。但下游所有实际消费方都额外要求 `quoteId`：

```ts
// TxFeeInfo.tsx:227
const isGasAccountEligible = gasAccountUiState.gasAccountEligible && !!gasAccountQuote?.quoteId;
// TxConfirmActions.tsx:206-208 / usePreCheckTokenBalance.ts:109-110
// ServiceSend.ts:259-260 广播 gate
gasAccountUiState?.selectedPayer === 'gasAccount' && gasAccountUiState.gasAccountQuote?.quoteId
```

`IGasAccountQuote.quoteId` 虽然类型上是必填 `string`，但它来自网络响应、运行时无保证，仓库里现有 4 处都做了 `?.quoteId` 防御，唯独新抽出的这个判定条件没有对齐。

## 影响

当后端返回 `gasAccountEligible=true` 且 `gasAccountQuote` 存在但 `quoteId` 为空串时，`resolveSponsorPayerState` 会同时给出 `effectiveFeePayer='gasAccount'` 和 `selectedPayer='gasAccount'`：UI 侧 `isGasAccountSponsored = effectiveFeePayer === 'gasAccount'` 为 `true`，`isPayerManagedByService` 隐藏手续费、展示赞助文案；同时 `idempotencyKey` 被 `buildGasAccountIdempotencyKey('')` 算成 `''`，广播时 `quoteId` 校验不通过、`quoteId`/`idempotencyKey` 都不会附带。

结果就是本 PR 最想避免的那类偏差：用户看到“已赞助 / 0 手续费”，实际按 user-paid 广播、自付 gas。Private Send 新增的 `serverPayer='megafuel'` 回落路径把这个窗口扩大到了此前 `effectiveFeePayer` 恒为 `'user'` 的场景。

## 建议

把判定条件与下游 gate 统一到 `quoteId` 上：

```ts
gasAccountQuoteEligible: !!(r.gasAccountEligible && r.gasAccountQuote?.quoteId),
```

同时建议在 `gasAccountPayerSelection.test.ts` 补一条用例，固定“quote 对象存在但 `quoteId` 为空 ⇒ `user`/`user`”这一不变量。

[Claude session ↗](https://claude.ai/code/session_01MkpLWw9pkfEVxChbKRG6Pt)

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by sidmorizon — COMMENTED** [review]
Review

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/kit/src/views/SignatureConfirm/utils/gasAccountPayerSelection.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+export function resolveSponsorPayerState({
+  serverPayer,
+  megafuelSponsorable,
+  gasAccountQuoteEligible,
+  isCustomRpcEnabled,
+  sponsorDisabledForBatch,
+  megafuelDisabledForPrivateSend,
+  gasAccountDisabledByScenario,
+  gasAccountTemporarilyDisabled,
+}: IResolveSponsorPayerStateParams): ISponsorPayerState {
```
Comment (sidmorizon) [inline packages/kit/src/views/SignatureConfirm/utils/gasAccountPayerSelection.ts:61]: **Bug-fix retrospective entry required by repo rules is missing**

Severity: `non-severe`

The repository's mandatory self-testing rule requires every bug fix to append a case entry to `.claude/skills/1k-retrospective/references/case-studies.md`, but this fix (`resolveSponsorPayerState` at `packages/kit/src/views/SignatureConfirm/utils/gasAccountPayerSelection.ts:34`) ships without such an entry.
Impact: The team loses the recorded lesson for this sponsorship-selection regression, so the same class of mistake is harder to catch next time.

File: packages/kit/src/views/SignatureConfirm/components/TxFee/TxFeeInfo.tsx (hunk 1 above)
Comment (weatherstar) [reply packages/kit/src/views/SignatureConfirm/components/TxFee/TxFeeInfo.tsx]: Fixed in 719c6aba3b. Extracted the eligibility check into `isGasAccountQuoteEligible()` (in `gasAccountPayerSelection.ts`), which requires `gasAccountEligible && gasAccountQuote?.quoteId` (non-empty), and switched the `TxFeeInfo.tsx` call site to it — so the `resolveSponsorPayerState` input is now aligned with every downstream `quoteId` gate (fee display, precheck, broadcast).

Also added unit tests fixing the invariant: a quote object with an empty `quoteId` is not eligible, and resolves to `user`/`user` even on the Private Send megafuel-fallback path where the window was widest.

File: packages/kit/src/views/SignatureConfirm/utils/gasAccountPayerSelection.ts (hunk 2 above)
Comment (weatherstar) [reply packages/kit/src/views/SignatureConfirm/utils/gasAccountPayerSelection.ts:61]: Fixed in 719c6aba3b. Appended the retrospective case entry ("Private Send dropped the Gas Account quote when the server preferred Megafuel") to `.claude/skills/1k-retrospective/references/case-studies.md`, covering the root cause (display payer and submit wiring derived from different inputs), the fix, and the review-caught `quoteId` guard.

**Comment by originalix** [comment]
## Automated code review found blocking issues

@weatherstar

Reviewed commit `719c6aba3b7e`. This report always reflects the latest reviewed changes.

### Review summary

This PR moves fee-payer display and Gas Account submit selection into one foreground SignatureConfirm decision boundary. It allows Private Send to fall back from suppressed Megafuel to an eligible Gas Account quote while preserving the existing custom-RPC, batch, and temporary-disable policies. The helper and decision-table tests improve maintainability, but the new quote-validity guard still does not control the display payer for a direct Gas Account response, leaving a user-visible fee mismatch.

**What needs attention:** Verify that every response lacking a usable Gas Account quote resolves both the displayed payer and submit payer to `user`, regardless of the server's raw `payer` value.

### Issues to address

- **P1 · Invalid Gas Account quotes still drive the sponsored UI**
  - **Impact:** When the service returns `payer='gasAccount'` with an empty `quoteId`, the new eligibility check makes `selectedPayer` fall back to `user`, but `effectiveFeePayer` remains `gasAccount`. `TxFeeInfo` consequently renders the sponsored zero-fee summary and hides the real network fee, while the submit path omits the unusable quote and broadcasts a user-paid transaction.
  - **Suggested action:** Make Gas Account quote eligibility gate `effectiveFeePayer` as well as `selectedPayer`, and change the existing no-eligible-quote test to require `user/user`.

### Validation gaps

- `git diff --check origin/x...HEAD` completed without whitespace errors.
- The targeted Jest suite could not run because `node_modules` is absent from this workspace.
- Existing GitHub review discussions could not be retrieved from the network, so duplicate coverage should be checked before posting the inline comment.
- No end-to-end environment was available to inject an id-less Gas Account response and confirm the displayed and charged fee.

File: packages/kit/src/views/SignatureConfirm/components/TxFee/TxFeeInfo.tsx (hunk 1 above)
Comment (sidmorizon) [reply packages/kit/src/views/SignatureConfirm/components/TxFee/TxFeeInfo.tsx]: `719c6aba3b` 只关掉了一半，直连 `serverPayer === 'gasAccount'` 这条路径上原缺陷仍在，建议先不要 resolve。

`isGasAccountQuoteEligible` 的抽取与 `TxFeeInfo.tsx:580` 的接入我已核对无误，`selectedPayer` 这一侧确实修好了。但 `effectiveFeePayer` 不经过该判定：

```ts
let effectiveFeePayer: IGasPayer = serverPayer;
if (
  isCustomRpcEnabled ||
  sponsorDisabledForBatch ||
  (gasAccountDisabledByScenario && serverPayer === 'gasAccount') ||
  (gasAccountTemporarilyDisabled && serverPayer === 'gasAccount')
) {
  effectiveFeePayer = 'user';
} else if (megafuelDisabledForPrivateSend && serverPayer === 'megafuel') {
  effectiveFeePayer = selectedPayer === 'gasAccount' ? 'gasAccount' : 'user';
}
```

取 `serverPayer = 'gasAccount'`、`quoteId` 为空、其余开关全 false：`gasAccountQuoteEligible` 为 `false` 使 `selectedPayer = 'user'`（正确），但四个重置条件都不命中，`else if` 又要求 `serverPayer === 'megafuel'`，于是 `effectiveFeePayer` 保持 `'gasAccount'`。

结果与本 thread 原本描述的后果完全一致：`isGasAccountSponsored` 为 `true`、`isPayerManagedByService` 隐藏真实网络费并显示赞助文案，而广播侧因 `selectedPayer === 'user'` 走 user-paid。只是抵达方式从「两侧都错判为 gasAccount」变成了「显示侧错、提交侧对」，用户看到的仍是「已赞助 / 0 手续费」却自付 gas。

新增用例没有覆盖到是因为它们都固定在 Private Send megafuel 回落场景（`serverPayer = 'megafuel'`），而这条路径的 `effectiveFeePayer` 恰好是从 `selectedPayer` 推导的，所以是对的。

建议让配额有效性同时门控两侧，例如在链首增加一条重置条件：

```ts
if (
  isCustomRpcEnabled ||
  sponsorDisabledForBatch ||
  (!gasAccountQuoteEligible && serverPayer === 'gasAccount') ||
  (gasAccountDisabledByScenario && serverPayer === 'gasAccount') ||
  (gasAccountTemporarilyDisabled && serverPayer === 'gasAccount')
) {
  effectiveFeePayer = 'user';
}
```

并补一条 `serverPayer: 'gasAccount'` + 空 `quoteId` ⇒ `user`/`user` 的用例——现有那条 no-eligible-quote 用例目前只断言了 `selectedPayer`。

[Claude session ↗](https://claude.ai/code/session_01MkpLWw9pkfEVxChbKRG6Pt)

---
_Generated by [Claude Code](https://claude.ai/code)_

Comment (weatherstar) [reply packages/kit/src/views/SignatureConfirm/components/TxFee/TxFeeInfo.tsx]: Fixed in f6b69f43c1. Added `(!gasAccountQuoteEligible && serverPayer === 'gasAccount')` as the first payer-specific reset condition in `resolveSponsorPayerState`, so the display side (`effectiveFeePayer`) is now gated by quote eligibility exactly like the submit side — the direct `serverPayer === 'gasAccount'` path with an empty `quoteId` now resolves to `user`/`user`.

Tests: updated the existing no-eligible-quote case to assert both payers reset to `user` (it previously pinned the buggy `effectiveFeePayer: 'gasAccount'` behavior), and added a dedicated end-to-end case `serverPayer: 'gasAccount'` + empty `quoteId` ⇒ `user`/`user` via `isGasAccountQuoteEligible`. All 21 tests pass.

**Review by originalix — APPROVED** [review]
